pkg/operator/controller/route-metrics/controller.go

```diff
@@ -0,0 +1,180 @@
+// that handles all the logic for gathering and exporting
+// metrics related to route resources.
+func New(mgr manager.Manager, namespace string) (controller.Controller, error) {
+	// Create a copy of the cluster, but with a new cache to watch on Route objects from every namespace.
+	newCluster, err := cluster.New(mgr.GetConfig(), func(o *cluster.Options) {
+		o.Scheme = mgr.GetScheme()
+	})
+	if err != nil {
+		return nil, err
+	}
+	// Add the cluster to the manager so that the cluster is started along with the other runnables.
+	mgr.Add(newCluster)
+	reconciler := &reconciler{
+		client:    mgr.GetClient(),
+		Namespace: namespace,
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	// add watch for changes in IngressController
+	if err := c.Watch(&source.Kind{Type: &operatorv1.IngressController{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, err
+	}
+	// add watch for changes in Route
+	if err := c.Watch(source.NewKindWithCache(&routev1.Route{}, newCluster.GetCache()),
+		handler.EnqueueRequestsFromMapFunc(reconciler.routeToIngressController)); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// routeToIngressController .
+func (r *reconciler) routeToIngressController(obj client.Object) []reconcile.Request {
+	var requests []reconcile.Request
+	// Cast the received object into Route object.
+	routeObject := obj.(*routev1.Route)
+
+	// Iterate through the related Ingress Controllers.
+	for _, ic := range routeObject.Status.Ingress {
+		log.Info("queueing ingresscontroller", "name", ic.RouterName)
+		// Create a reconcile.Request for the Ingress Controller.
+		request := reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      ic.RouterName,
+				Namespace: r.Namespace,
+			},
+		}
+		requests = append(requests, request)
+	}
+	return requests
+}
+
+// reconciler handles the actual ingresscontroller reconciliation logic in response to events.
+type reconciler struct {
+	client    client.Client
+	Namespace string
+}
+
+// Reconcile expects request to refer to a Ingress Controller resource, and will do all the work to gather metrics related to
+// the resource.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log.Info("reconciling", "request", request)
+
+	// Fetch the Ingress Controller object.
+	ingressController := &operatorv1.IngressController{}
+	if err := r.client.Get(ctx, request.NamespacedName, ingressController); err != nil {
+		if kerrors.IsNotFound(err) {
+			// This means the Ingress Controller object was already deleted/finalized.
+			log.Info("Ingress Controller not found", "request", request)
+			// Delete the Shard label corresponding to the Ingress Controller from the RouteMetricsControllerRoutesPerShard metric.
+			DeleteRouteMetricsControllerRoutesPerShardMetric(request.Name)
```

Comment:
Perhaps we should not delete metrics.  What if the ingressController is being recreated?

Reply:
The delete metrics would be required or else stale metrics would still get exported. Additionally, this will effect the telemetry data that would be sent to Red Hat server.

If the Ingress Controller object is recreated, then that event would trigger another reconcile. This reconcile will take care of adding back the correct metric labels.